### PR TITLE
[Refactor] Refactoring Expert Parallelism

### PR DESCRIPTION
**Pull request by NouamaneTazi**

### CODE REVIEW

File: src/nanotron/parallel/sharded_parameters.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ def create_sharded_parameter_from_config(
 ) -> NanotronParameter:
     current_rank = dist.get_rank(pg)
     param_num_dims = len(parameter.shape)
-    global_ranks = tuple(sorted((dist.get_global_rank(pg, i) for i in range(pg.size()))))
+    global_ranks = dist.get_global_ranks(pg)
```
Comment (3outeille) [inline src/nanotron/parallel/sharded_parameters.py:43]: do we still need it to be sorted as previously ?

**Comment by 3outeille** [comment]
@NouamaneTazi left some comments before merging. Great job !

File: examples/moe/moe.py [hunk 2]
```diff
@@ -264,7 +271,7 @@ def parallel_forward_once(self, x, expert_weights, top_experts):
             parallel_bin_ids,
             None,  # expert_weights
             parallel_bins,
-            num_experts_per_tok=self.num_experts_per_tok,
+            num_experts_per_tok=1,
```
[the remark this answers is not in the document]
Comment (NouamaneTazi) [reply examples/moe/moe.py:274]: yes 
https://github.com/stanford-futuredata/megablocks/blob/014667c87df638dc048e529e09708053e5010685/megablocks/layers/moe.py#L393

File: examples/moe/moe.py [hunk 3]
```diff
@@ -126,7 +129,11 @@ def __init__(
         self.forward_fn = self.parallel_forward_once if self.expert_pg_size > 1 else self.forward_once
 
         self.blocking = 128
-        self.mlp = MLP(config=config, parallel_config=parallel_config, tp_pg=tp_pg)
+
+        if self.experts_per_rank == 1:
+            self.mlp = MLP(config=config, parallel_config=parallel_config, tp_pg=parallel_context.tp_pg)
```
[the remark this answers is not in the document]
Comment (NouamaneTazi) [reply examples/moe/moe.py:134]: This case still need more refactoring actually which is why I ignored it for now. I'm planning to remove this case in a next PR 🙌

**Pull request merged**